### DISCUSSION TITLE
How do I pass params to a custom typst template

### DISCUSSION THREAD

**Discussion by Canadauni** [quarto-dev/quarto-cli#10255] [Q&A] [answered]
### Description

Hey There,

I'm working with a custom Typst Format that I developed to generate a series of static pdf dashboards (essentially a 1 page report) and I'd like to set the header of the document from the `quarto_render()` function call. I have WIP version of the template on my profile but hopefully the below will be a sufficient repeat reproducible example.

typst_template.typ:
```typ
#let dashboard(
  title: "Dashboard Title"
) = {
  set page(
    header: text(weight: "bold", title)
  )
  body
}
````

typst_show.typ:
````typ
#show: dashboard.with(
$if(title)$
  title: [$title$],
$endif$
)
````

quarto doc:
````qmd
---
format:
  typst_static_dashboard-typst: default
params:
  title: "overwrite params"
title: "overwrite metadata"
---
REST OF DOC
````

In the code I use to batch run the data I use
```r
quarto_render(
  input = "dashboard_pdf.qmd",
  output_file = "output.pdf",
  metadata = list(title = "Site Name")
)
```

My (perhaps wrong) expectation is that header will show my updated title "Site Name" instead of "overwrite metadata" but the output pdf shows the "overwrite metadata". Interestingly the metadata in the log does show `title: Site Name`.

I've also tried setting it via params with the following typst_show and the `execute_params` arg instead of the `metadata` arg in the function call with the same results:
typst_show.typ:
````typ
#show: dashboard.with(
$if(params.title)$
  title: [$params.title$],
$endif$
)
````

I was wondering if this was a bug or more likely I'm missing something about the rendering pipeline here but would like to confirm!

Thanks!

**Comment by cscheid** [comment]
That general strategy looks right: add metadata to document, and then use it in the template within `$ ... $` blocks.

One documentation issue that we need to address is a complete description of which metadata fields are "safe" for third-party templates to use. I think that using `params` is a bit chancy because it's the entry under which the `knitr` engine does parameterized documents, and it's possible that Quarto is inadvertently mangling that.

Can you share a git repository with your attempt so I can take a look at it? Thanks.

**Reply by Canadauni** [reply]
This is the repo for my template: https://github.com/Canadauni/typst-static-dashboard

I'm still figuring out the process, it won't have my rendering pipeline and it's missing some of the troubleshooting I've being doing, but what I've put in the example is pretty much what I've been doing.

**Reply by mcanouil** [reply]
To clarify what metadata and what parameters are.
Parameters are for computations inside the engine such as `knitr`, Jupyter or Julia.
In the R package, the argument is the same as in Quarto CLI namely "execute_params".
"metadata" controls metadata (no computations), see the R package documentation https://quarto-dev.github.io/quarto-r/reference/quarto_render.html (or `quarto render --help`)

Because you are not using Quarto CLI directly, you add another layer where there could be a bug.
I suggest you try using Quarto CLI directly:
```sh
quarto render dashboard_pdf.qmd --output output.pdf --metadata "title:Site Name"
```

Below, an example using Python and "metadata-file" to update metadata used in a Quarto extension for a Typst format:
https://github.com/mcanouil/generate-quarto-invoices


---

FYI: `labs(x = "", y = "")` means have an empty X/Y axis titles. It does not mean "no x/y axis title, `labs(x = NULL, y = NULL)` does.

**Reply by mcanouil** [reply]
I made a PR to your repository:
- Canadauni/typst-static-dashboard#1

<table>
<tr><th>Input</th><th>Output</th></tr>
<tr><td>

````sh
quarto render template.qmd --metadata "title:Site Name"
````

</td><td>

<img width="924" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/a53c6304-42f8-408a-a354-3426daf30a02">

</td></tr>
</table>

Note that instead of relying on computation which require to have R/knitr setup, you could use the placeholder and lipsum shortcodes from Quarto 1.5.
https://quarto.org/docs/authoring/shortcodes.html

<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/45b91d98-1f04-4c7c-84e6-e90c7f50c58f">

**Reply by cscheid** [reply]
Thanks for the repo - I don't see the equivalent snippet of source code you shared here, so I don't know what you consider to not be working at this point. If your problem is solved, that's great! If not, can you let me know by referencing things in that repo? Thanks.

**Reply by Canadauni** [reply]
Thanks both of you for looking into things, I think the template is working as expected and returning to my question this morning I think I perhaps could have phrased the question/problem a bit better. I'm having more of an issue with the render pipeline and I think @mcanouil got to it by suggesting I use:

```
quarto render dashboard_pdf.qmd --output output.pdf --metadata "title:Site Name"
```
which results in: 
![image](https://github.com/quarto-dev/quarto-cli/assets/19292113/c5a2539b-4155-4df5-907a-3a34bfff98c8)

However since I'm trying to batch process a bunch of data and generate reports I'd like to use:
```r
quarto::quarto_render(
  input = "dashboard_pdf.qmd",
  output_file = "output.pdf",
  metadata = list(title = "Site Name")
)
```
which results in:
![image](https://github.com/quarto-dev/quarto-cli/assets/19292113/a2ec6cb2-724c-4453-827f-3e51c2923cba)

Looking at the logs from each, it looks like there is a difference in what is passed to pandoc.

Using the the CLI the result is:
```
processing file: dashboard_pdf.qmd

output file: dashboard_pdf.knit.md

pandoc --output dashboard_pdf.typ --metadata title:Site Name
  to: typst
  standalone: true
  shift-heading-level-by: -1
  default-image-extension: svg
  wrap: none
  citeproc: false

metadata
  title: Static Dashboard Template
  grid-cols: 3
  grid-rows: 3
  paper: us-letter
  flipped: true
  header-col: FFA500
  font-size: 10pt
  font-col: FFA500

[typst]: Compiling dashboard_pdf.typ to dashboard_pdf.pdf...DONE

Output created: output.pdf
```

However, using the R function I get:
```
processing file: dashboard_pdf.qmd

output file: dashboard_pdf.knit.md

pandoc --output dashboard_pdf.typ
  to: typst
  standalone: true
  shift-heading-level-by: -1
  default-image-extension: svg
  wrap: none
  citeproc: false

metadata
  title: Site Name
  grid-cols: 3
  grid-rows: 3
  paper: us-letter
  flipped: true
  header-col: FFA500
  font-size: 10pt
  font-col: FFA500

[typst]: Compiling dashboard_pdf.typ to dashboard_pdf.pdf...DONE

Output created: output.pdf
```

**Reply by Canadauni** [reply, marked as answer]
Oh D'Oh I just checked the R project and found this issue: https://github.com/quarto-dev/quarto-r/issues/181 which is my exact problem.

Sorry for taking your time on this, I do appreciate all the help and the PR!

**Reply by mcanouil** [reply]
So this seems to be an issue with the R package wrapper.

**Reply by Canadauni** [reply]
Yah that looks to be the case, thanks again for the help troubleshooting this.
